### PR TITLE
cattle example, LD calculator cleaned up

### PR DESCRIPTION
**Pull request by ningyuxin1999**
This is the example of cattle configuration. Each population size is dependent on the previous one instead of generated randomly.

### CODE REVIEW

**Comment by nspope** [comment]
Great, thanks Yuxin! Could you please rebase this onto main and fix the merge conflicts?

If you haven't done this before, then from the command line
```bash
git fetch upstream main  # this is assuming that this repo is called "upstream" in `git remote -v`
git checkout processor_cleanup
git rebase main
```
and then you'll have to edit the files that git lists, to resolve the 
```bash
<<<<< HEAD
old code
=======
new code
>>>>>> commit
```
inserts added by git. Let me know if you have questions!

**Comment by andrewkern** [comment]
hmm this still has merge conflicts @ningyuxin1999

**Comment by ningyuxin1999** [comment]
> hmm this still has merge conflicts @ningyuxin1999

I was only updating some comments yesterday, and now the conflicts should be resolved :)

**Review by nspope — CHANGES_REQUESTED** [review]
Thanks @ningyuxin1999 -- it looks like there's a few things to fix that I've flagged here.

Also, it'd be good if you could try running the training pipeline through on this example just to make sure it doesn't error out (the number of sims can be set to something low, like 100, just want to make sure it doesn't error out).

@andrewkern can you take a quick look as well?

File: workflow/scripts/ts_simulators.py [hunk 1]
```diff
@@ -256,6 +256,138 @@ def __call__(self, seed: int = None) -> (tskit.TreeSequence, np.ndarray):
...
+        corresponding to the most recent time window. The following population
+        sizes are generated following N_i = N_{i-1} * 10 ^ β for i in [1,21].
+        """
+        # Sample the first value uniformly within the log10 bounds
+        log_values = self.prior.sample().numpy()
+        # For subsequent time windows, sample a new value based on the previous one
+        for i in range(1, self.num_time_windows):
+            new_value = 10 ** self.pop_sizes[0] - 1
+            while new_value > self.pop_sizes[1] or new_value < self.pop_sizes[0]:
+                new_value = log_values[i - 1] * 10 ** np.random.uniform(-1, 1)
```
[the remark this answers is not in the document]
Comment (nspope) [reply workflow/scripts/ts_simulators.py:320]: once I understand what the intended process is here, maybe we could reparameterize in terms of iid uniforms (e.g. without rejection sampling), as this is most definitely not a box uniform any more

File: workflow/config/cattle_21Gen.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+optimizer: "Adam"
+batch_size: 32
+learning_rate: 0.0005
+max_num_epochs: 200
+stop_after_epochs: 20
+clip_max_norm: 5
+packed_sequence: False
+
+simulator:
+  class_name: "VariablePopulationSize"
```
[the remark this answers is not in the document]
Comment (andrewkern) [reply workflow/config/cattle_21Gen.yaml:34]: yep, agreed

File: workflow/scripts/ts_processors.py [hunk 3]
```diff
@@ -198,110 +198,63 @@ class tskit_windowed_sfs_plus_ld(BaseProcessor):
...
-        distance_bins : int or list LD will be averaged by bins of distances e.g. if distance_bins = [0, 100, 1000,
-        10000], LD will be averaged for the groups [0,100[, [100, 1000[, and [1000, 10000[ If distance_bins is an int,
-        it defines the number of bins of distances for which to compute the LD The bins are created in a logspace If
-        distance_bins is a list, they will be used instead
-
-        Returns
-        -------
-        DataFrame
-            Table with the distance_bins as index, and the mean value of
+        Compute LD for a subset of SNPs and return a DataFrame with only the mean r2.
```
Comment (andrewkern) [inline workflow/scripts/ts_processors.py:204]: i think it's useful to keep some of the details about the parameters that are needed in here

File: workflow/scripts/ts_processors.py [hunk 4]
```diff
@@ -198,110 +198,63 @@ class tskit_windowed_sfs_plus_ld(BaseProcessor):
...
             snps = np.arange(0, n_SNP, gap) + np.random.randint(0, (n_SNP - 1) % gap + 1)
-            # adding a random start (+1, bc 2nd bound in randint is exlusive)
-
-            # non overlapping contiguous pairs
-            # snps=[ 196, 1220, 2244] becomes
-            # snp_pairs=[(196, 1220), (1221, 2245)]
-            snp_pairs = np.unique([((snps[i] + i) % n_SNP, (snps[i + 1] + i) % n_SNP) for i in range(len(snps) - 1)],
-                                  axis=0)
+            snp_pairs = np.unique([((snps[i] + i) % n_SNP, (snps[i + 1] + i) % n_SNP)
+                                for i in range(len(snps) - 1)], axis=0)
```
Comment (andrewkern) [inline workflow/scripts/ts_processors.py:222]: for my eyes, it's way easier to have the list comprehension on a single line, splitting list comprehensions across lines makes things harder to read

File: workflow/scripts/ts_processors.py [hunk 5]
```diff
@@ -198,110 +198,63 @@ class tskit_windowed_sfs_plus_ld(BaseProcessor):
...
-
+            sd = pd.DataFrame((np.diff(pos_vec[snps_pos])), columns=["snp_dist"])
             sd["dist_group"] = pd.cut(sd.snp_dist, bins=distance_bins)
             sr = [allel.rogers_huff_r(snps) ** 2 for snps in haplotype[snps_pos]]
             sd["r2"] = sr
             sd["gap_id"] = i
             ld = pd.concat([ld, sd])
 
-        ld2 = ld.dropna().groupby("dist_group", observed=False).agg(agg_bins)
+        ld2 = ld.dropna().groupby("dist_group",observed=True).agg(agg_bins)
```
Comment (andrewkern) [inline workflow/scripts/ts_processors.py:248]: i believe that setting `observed=True` here will throw an error, at least it did for me when first reimplementing this routine

File: workflow/scripts/ts_simulators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewkern) [reply workflow/scripts/ts_simulators.py:320]: i'm a bit confused here too because the vector named `log_values` will initially be populated by the BoxUniform draws

**Comment by andrewkern** [comment]
added some additional comments

File: workflow/scripts/ts_simulators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py:320]: I think `log_values` is confusing here. What this should actually do (and is currently not) is drawing the first population size between 10 and 100000 but uniformly in log10-scale. So if we draw alpha from a BoxUniform between 1 and 5 the first population size should then be 10^alpha. In each next time window the population size is then N_{i+1} = N_i * 10^beta where beta is uniform in [-1,1].

[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py:320]: This is to prevent the posterior distribution from focusing too much on the larger population sizes and makes thus probably sense also for the variablepopulationsizes class. I think the performance for the more realistic population size scenarios should also improve using this prior with dependent population sizes as it prevents completely weird population size changes. Basically if we would draw the population sizes uniformly and independently between 10 and 100000 too many of the training data would contain large ancestral population sizes.

[the remark this answers is not in the document]
Comment (nspope) [reply workflow/scripts/ts_simulators.py:320]: This is a bit of a tangent, but worth discussing ... I think we should try to have keep the posterior and prior parameterization consistent. For example (simplifying things for exposition), say I have a model parameterized by the random walk `Ne[i] = Ne[i - 1] * U[lower, upper]`. Then the values of theta that the simulator returns -- the targets for the NN -- should be the `U[lower, upper]` values, not the Ne values, because we're assuming a uniform prior downstream.

[the remark this answers is not in the document]
Comment (nspope) [reply workflow/scripts/ts_simulators.py:320]: Also I think using np.random.uniform here won't respect the random seed, as this has only been set for torch

[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py:320]: Ok, we could certainly use the U[lower, upper] values instead of the actual population sizes. But we would have to change a small detail in the generation of the population sizes then. Currently, we redraw U[lower, upper] if Ne[i-1] * U[lower, upper] is larger than the maximal population size allowed (or smaller than the minimal). Instead of redrawing we could simply set the population size to the max or min value allowed in this case.

File: workflow/config/cattle_21Gen.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  # OPTIONAL CLASS KWARGS:
+  sequence_length: 2000000
+  mutation_rate: 1e-8
+  num_time_windows: 21
+  max_time: 130000
+  time_rate: 0.06
+  samples:
+    pop: 25
```
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/config/cattle_21Gen.yaml]: Done

File: workflow/config/cattle_21Gen.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/config/cattle_21Gen.yaml]: I changed this, but in the long run I think I would like to use the same class for the cattle example and the VariablePopulationSize example. In my opinion the dependent population sizes make much more sense as a prior and could be used in any case.

File: workflow/scripts/ts_simulators.py [hunk 7]
```diff
@@ -256,6 +256,138 @@ def __call__(self, seed: int = None) -> (tskit.TreeSequence, np.ndarray):
...
+            num_sample = geno.shape[0]
+            if (geno==2).any():
+                num_sample *= 2
+            
+            row_sum = np.sum(geno, axis=0)
+            keep = np.logical_and.reduce([
+                row_sum != 0,
+                row_sum != num_sample,
+                row_sum > num_sample * 0.05,
+                num_sample - row_sum > num_sample * 0.05
```
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py]: Done

File: workflow/scripts/ts_simulators.py [hunk 8]
```diff
@@ -256,6 +256,138 @@ def __call__(self, seed: int = None) -> (tskit.TreeSequence, np.ndarray):
...
+            # Check if we have enough SNPs after MAF filtering
+            geno = ts.genotype_matrix().T
+            num_sample = geno.shape[0]
+            if (geno==2).any():
+                num_sample *= 2
+            
+            row_sum = np.sum(geno, axis=0)
+            keep = np.logical_and.reduce([
+                row_sum != 0,
+                row_sum != num_sample,
```
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py]: Yes, I deleted them.

File: workflow/scripts/ts_simulators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbaumdicker) [reply workflow/scripts/ts_simulators.py]: I have now implemented the new version, where the population sizes will be a bit more sticky at the population size range borders.

**Comment by fbaumdicker** [comment]
I started to fix the issues in this branch. Currently, I am no longer including the LD statistics and am just using the sfs. I will look at the LD stuff again later. The new prior is done and seems to work fine.
However, when using the cattle config I currently get stuck in the plot_diagnostics job when drawing the posterior samples. 
`snakemake --configfile workflow/config/cattle_21Gen.yaml --snakefile workflow/training_workflow.smk`
On my PC I run out of GPU memory, on the workstation, drawing from the prior does not progress. Any idea what could cause this @nspope ? I want to fix this first before continuing with the LD calculation part.

**Comment by nspope** [comment]
Hm ... we're batching the sampling during plot diagnostics, so the per batch memory requirement should be pretty tightly constrained (as much as it would be during training). But maybe I need to be transferring results to the CPU at the end of each batch. Any chance you could drop some print statements into `workflow/scripts/plot_diagnostics.py` to try to figure out if the OOM is happening within Lightning's prediction loop, or afterwards?

I'm not sure what could be happening with the prior -- do you mean that it's stuck in an endless while loop during the simulation?

**Comment by andrewkern** [comment]
i've seen this behavior once before with the posterior sampling being stuck in an infinite loop. it had to do with not properly defining the prior for use with SBI. looking at your code i don't see anything obvious, but i'd go back and compare it to the `VariablePopnSize` simulator [here](https://github.com/kr-colab/popgensbi_snakemake/blob/e6d91fee1e8b6dac1a0c1adbb30aafb3011f4ef7/workflow/scripts/ts_simulators.py#L146C45-L147C1) (which works) to see if there is something off

**Comment by fbaumdicker** [comment]
Thanks for the hint @andrewkern. Took me a while, but I found the bug. I was passing on the transformed values instead of the prior samples. However, I learned a lot while searching for this and found another unrelated bug.
I am also no longer convinced of the current parametrization of the dependent population sizes using the population size changes instead of the actual population sizes. Mainly due to the harder-to-calculate resulting posterior distribution for the actual population sizes. I will think a bit more about this.
Maybe we can define a custom prior that uses dependent sampling but provides the logprob as if it were uniformly sampled?

**Comment by nspope** [comment]
> Maybe we can define a custom prior that uses dependent sampling but provides the logprob as if it were uniformly sampled?

Sorry I'm not following -- do you mean, have the stored prior be uniform but the actual prior be something more complex, with the same support?

**Comment by fbaumdicker** [comment]
Yes, that is what I was thinking of. As long as the support is the same this should in principle work, but I do not know how this would affect the predictions. So maybe this is not the route to go?
Actually, it wouldn't be too hard to compute the correct logprob for the scenario I have in mind. However, it would be a custom prior and this is currently not included in the snakemake workflows. So if this would not be too hard to include, this would be my preferred path. We have defined custom priors in the past, so this should not be a big issue as long as we can import the prior to all subsequent steps when necessary. For example the plot_diagnostics script currently assumes BoxUniform as a prior.

**Comment by nspope** [comment]
I don't think it's an issue with what's in there now; as we're using the prior samples (not the log density) for everything currently implemented. Where it gets a little hairy, conceptually, is when we try to aggregate posteriors across windows. (Because then we effectively duplicate the prior). But maybe we just accept that the aggregation is not sensible to use with a non-uniform prior.

**Comment by andrewkern** [comment]
checking in on this PR-- @fbaumdicker what's the current status here?

**Comment by fbaumdicker** [comment]
We fixed the (hopefully) last bug today. So I am now cleaning up the code and will then rebase.
We now have a simulator example with a working prior that is not a BoxUniform and where the parameters sampled from the prior depend on each other.  We are currently running a comparison with your previous version with a BoxUniform prior with independent parameters on some example datasets. This should give a nice example for the usage of more complex priors. But this PR should be independent of this analysis and be ready to merge soon.

**Comment by nspope** [comment]
Hey @ningyuxin1999, thanks for putting up the code. As Andy says, we want to keep things as compartmentalized as possible to avoid impacting the other experiments. I think the easiest thing to do here would be to make a new PR, where you do the following:

1. Make a new directory `experiments/variablepopsize-dependent-prior` with a subdirectory `npe-config`
2. Copy the altered `config/variablepopsize.yaml` that you've made in this PR to `experiments/variablepopsize-dependent-prior/npe-config/variablepopsize.yaml`. (Note that Andy pointed to some odd resource usage, like requesting 20 GPUs, this is probably a typo)
3. Copy over the new `DependentVariablePopulationSize` class that you've added in this PR into the new PR, in the same location (`ts_simulators.py`). Please don't edit the other `VariablePopulationSize` class as this is what Andy was using for his experiments.
4. Copy over the changes you've made to `tskit_windowed_sfs_plus_ld` class in this PR into the new PR (in the same location, `ts_processors.py`).
5. Add a shell script to `experiments/variablepopsize-dependent-prior/run-experiment.sh` that gives the snakemake command to train the model (e.g. using the config in `experiments/variablepopsize-dependent-prior/npe-config`)
6. Put any code to generate figs into python scripts in `experiments/variablepopsize-dependent-prior`

Then we'll have the code organized in the same way as for the other parts of the paper, and there shouldn't be edits to classes other than those that are unique to this experiment. The commit history will also be a bit cleaner as there won't be a merge from main into this feature branch. Let me know if you have questions or if you want to talk through the code organization. Thanks!!!

**Comment by ningyuxin1999** [comment]
Thanks for your reviews and suggestions! @andrewkern @nspope 🙏 I will open the new PR tomorrow or latest on Wednesday with the dependent prior and the new experiments using that prior.
